types/src/key.rs

```diff
@@ -58,11 +67,442 @@ impl From<HashAddr> for Key {
+    }
+}
+
+impl FromBytes for ContractWasmHash {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (bytes, rem) = FromBytes::from_bytes(bytes)?;
+        Ok((ContractWasmHash::new(bytes), rem))
+    }
+}
+
+impl From<[u8; 32]> for ContractWasmHash {
+    fn from(bytes: [u8; 32]) -> Self {
+        ContractWasmHash(bytes)
+    }
+}
+
+impl Serialize for ContractWasmHash {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            self.to_formatted_string().serialize(serializer)
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for ContractWasmHash {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        if deserializer.is_human_readable() {
+            let formatted_string = String::deserialize(deserializer)?;
+            ContractWasmHash::from_formatted_str(&formatted_string).map_err(SerdeError::custom)
+        } else {
+            let bytes = ContractWasmHashBytes::deserialize(deserializer)?;
+            Ok(ContractWasmHash(bytes))
+        }
+    }
+}
+
+impl AsRef<[u8]> for ContractWasmHash {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<&[u8]> for ContractWasmHash {
+    type Error = TryFromSliceForContractHashError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, TryFromSliceForContractHashError> {
+        ContractWasmHashBytes::try_from(bytes)
+            .map(ContractWasmHash::new)
+            .map_err(|_| TryFromSliceForContractHashError(()))
+    }
+}
+
+impl TryFrom<&alloc::vec::Vec<u8>> for ContractWasmHash {
+    type Error = TryFromSliceForContractHashError;
+
+    fn try_from(bytes: &Vec<u8>) -> Result<Self, Self::Error> {
+        ContractWasmHashBytes::try_from(bytes as &[u8])
+            .map(ContractWasmHash::new)
+            .map_err(|_| TryFromSliceForContractHashError(()))
+    }
+}
+
+#[cfg(feature = "std")]
+impl JsonSchema for ContractWasmHash {
+    fn schema_name() -> String {
+        String::from("ContractWasmHash")
+    }
+
+    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
+        let schema = gen.subschema_for::<String>();
+        let mut schema_object = schema.into_object();
+        schema_object.metadata().description =
+            Some("The hash address of the contract wasm".to_string());
+        schema_object.into()
+    }
+}
+
 /// An alias for [`Key`]s hash variant.
-pub type ContractPackageHash = HashAddr;
+pub type ContractPackageHashBytes = HashAddr;
+
+/// A newtype wrapping a `ContractPackageHashBytes` which is the raw bytes of
+/// the ContractPackageHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractPackageHash(ContractPackageHashBytes);
+
+impl ContractPackageHash {
+    /// Constructs a new `ContractHash` from the raw bytes of the contract hash.
```

Comment:
```suggestion
    /// Constructs a new `ContractPackageHash` from the raw bytes of the contract hash.
```

Also, as above, please change "contract hash" to "contract package hash" throughout comments.

Reply:
Changed comments